promql/functions.go

```diff
@@ -75,26 +77,31 @@ func extrapolatedRate(vals []parser.Value, args parser.Expressions, enh *EvalNod
 		resultHistogram    *histogram.FloatHistogram
 		firstT, lastT      int64
 		numSamplesMinusOne int
+		annos              = annotations.Annotations{}
 	)
 
 	// We need either at least two Histograms and no Floats, or at least two
 	// Floats and no Histograms to calculate a rate. Otherwise, drop this
 	// Vector element.
+	metricName := samples.Metric.Get(labels.MetricName)
 	if len(samples.Histograms) > 0 && len(samples.Floats) > 0 {
-		// Mix of histograms and floats. TODO(beorn7): Communicate this failure reason.
-		return enh.Out
+		return enh.Out, annos.Add(annotations.NewMixedFloatsHistogramsWarning(metricName, args[0].PositionRange()))
+	}
+
+	if isCounter && !strings.HasSuffix(metricName, "_total") && !strings.HasSuffix(metricName, "_sum") && !strings.HasSuffix(metricName, "_count") {
```

Comment:
I think you should add `metricName==""` as a case that is not warned.
This would happen for instance if the input to `rate` comes from a calculation, and not from a series.

Reply:
Yeah... but is there a valid scenario where that can happen? Rates should always be done on raw counters coming from the storage before any other operation is done on them... to pass in a derived expression into `rate()` you would have to do a subquery to turn it into a range vector, and I'm not sure there's ever a case where it makes sense to do that for counters before rate-ing them?

Reply:
second derivative?

Reply:
Hmpf. Ok fine :)